### ISSUE TITLE
Pub command to deploy a web dir

### ISSUE THREAD

**Issue by DartBot** [dart-lang/pub#221]
<a href="https://github.com/sethladd"><img src="https://avatars.githubusercontent.com/u/5479?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [sethladd](https://github.com/sethladd)**
_Originally opened as dart-lang/sdk#6006_

----

I need to deploy a web/ dir that's inside my pub package. It would be great if pub had some command to make this easy. Perhaps, create a tar.gz of web/ with the correct packages/ copy or symlink.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Added this to the [Later](../milestones/Later) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Removed this from the [Later](../milestones/Later) milestone.*
*Added this to the [M3](../milestones/M3) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Marked this as blocking dart-lang/sdk#8263.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

Issue dart-lang/sdk#8102 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Removed this from the [M3](../milestones/M3) milestone.*
*Added this to the [M4](../milestones/M4) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Removed this from the [M4](../milestones/M4) milestone.*
*Added this to the [M5](../milestones/M5) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

*Set owner to @nex3.*
*Added [Started](../labels/Started) label.*

**Comment by DartBot** [comment]
<a href="https://github.com/sethladd"><img src="https://avatars.githubusercontent.com/u/5479?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sethladd](https://github.com/sethladd)**

----

Can you give us an idea of what this will do? Thanks!

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

The immediate goal is just to bake all the external dependencies in to a directory with no outgoing symlinks. This directory can then be (relatively) easily copied to a server and served statically, or run locally without fear of being disrupted by e.g. deletion of the pub cache.

**Comment by DartBot** [comment]
<a href="https://github.com/sethladd"><img src="https://avatars.githubusercontent.com/u/5479?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sethladd](https://github.com/sethladd)**

----

Will it handle static assets that come in from a package? e.g. dart.js ?  (not sure what you mean by "external dependencies")

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

We've talked this over a little more, and we have a more specific idea of what it's going to do. It'll copy everything in "web" into a deploy directory, run dart2dart and dart2js on all the entrypoints, and clean up the Dart source files. This will produce a directory that can be served without any additional steps.

**Comment by DartBot** [comment]
<a href="https://github.com/sethladd"><img src="https://avatars.githubusercontent.com/u/5479?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sethladd](https://github.com/sethladd)**

----

By copy, you mean "if it's symlinked in, it'll get actually copied to this build directory".  So, to get 3rd party assets like dart.js, just make sure they are in their own lib directory. ?

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

The deploy directory won't have any package symlinks. All the Dart code will be compiled together using dart2dart or dart2js.

For the first iteration of this, we don't plan to support assets.

**Comment by DartBot** [comment]
<a href="https://github.com/sethladd"><img src="https://avatars.githubusercontent.com/u/5479?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sethladd](https://github.com/sethladd)**

----

Thanks, I understand the deploy dir won't have symlinks. Was wondering how dart.js will get into my deploy dir. Sounds like I can add a manual copy step to get that and other static assets into deploy dir.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Pub will need to access dart2js, but currently the relative path from pub to dart2js differs between the repo and the SDK.
______
*Marked this as being blocked by dart-lang/sdk#6619.*

**Comment by DartBot** [comment]
<a href="https://github.com/dgrove"><img src="https://avatars.githubusercontent.com/u/2108507?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [dgrove](https://github.com/dgrove)**

----

Issue dart-lang/sdk#10544 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

Issue dart-lang/sdk#9996 has been merged into this issue.
______
 cc @dgrove.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Added [Fixed](../labels/Fixed) label.*

**Comment by DartBot** [comment]
*This comment was originally written by mdem...&#064;google.com*
_____
Any details on why this is "fixed"?  I assume there's a "pub deploy" command or something now, but I'm not having any luck Googling for details about it.  Also, http://pub.dartlang.org/doc/ including the "Docs" drop down menu don't mention any deploy commands.
